README.md
Leanify [![GitHub Actions CI](https://github.com/JayXon/Leanify/actions/workflows/ci.yml/badge.svg)](https://github.com/JayXon/Leanify/actions) [![Windows Build status](https://ci.appveyor.com/api/projects/status/2p9i3ru8apwq2uic?svg=true)](https://ci.appveyor.com/project/JayXon/leanify) [![Download](https://img.shields.io/github/downloads/JayXon/Leanify/total.svg)](https://github.com/JayXon/Leanify/releases) [![Latest Release](https://img.shields.io/github/release/JayXon/Leanify.svg)](https://github.com/JayXon/Leanify/releases/latest) [![GitHub license](https://img.shields.io/github/license/JayXon/Leanify.svg)](LICENSE)
=======

Leanify is a lightweight lossless file minifier/optimizer. It removes unnecessary data (debug information, comments, metadata, etc.) and recompress the file to reduce file size. It will not reduce image quality at all.


## Features

* Support recursive minifying. (e.g. a [PNG] inside an [APK] inside a [ZIP])
* Support a wide variety of [file formats](#file-formats).
* Lightweight, one file, under 1MB, no external dependency.
* Everything is done in memory, no temporary files.
* Cross-Platform, support Windows, Linux, Mac.
* Support traverse directory recursively.
* Ability to identify file format by its data instead of name.


## Disclaimer

I'm not respossible for any consequence of using Leanify.

**PLEASE BACKUP THE FILE BEFORE USING LEANIFY!**


## File Formats


#### APK file (.apk)

It is based on [ZIP].

Note that modifying files inside `APK` will break digital signature.
To install it, you'll have to sign it again.

If you don't want to modify any files inside `APK`, use `-d 1` option.


#### Comic book archive (.cbt, .cbz)

`cbt` is based on [tar]. `cbz` is based on [ZIP].


#### Microsoft Office document 2007-2013 (.docx, .xlsx, .pptx)

It is based on [XML] and [ZIP].

Office document 1997-2003 (.doc, .xls, .ppt) is not supported.


#### Data URI (.html .htm .js .css)

Looks for `data:image/*;base64` and leanify base64 encoded embedded image.


#### Design Web Format (.dwf, dwfx)

It is based on [ZIP].


#### EPUB file (.epub)

It is based on [ZIP].


#### FictionBook (.fb2, .fb2.zip)

It is based on [XML].

Leanify embedded images.


#### GFT file (.gft)

It's an image container format found in Tencent QQ.

Leanify the image inside.


#### gzip file (.gz, .tgz)

Leanify file inside and recompress deflate stream.

Remove all optional section: `FEXTRA`, `FNAME`, `FCOMMENT`, `FHCRC`.


#### Icon file (.ico)

Convert 256x256 BMP to [PNG].

Leanify [PNG] inside, if any.


#### Java archive (.jar)

It is based on [ZIP].


#### JPEG image (.jpeg, .jpg, .jpe, .jif, .jfif, .jfi, .thm)

Remove all application markers (e.g. `Exif` (use `--keep-exif` to keep it), `ICC profile`, `XMP`) and comments.

Optimize with `mozjpeg`.


#### Lua object file (.lua, .luac)

Remove all debugging information:

* Source name
* Line defined and last line defined
* Source line position list
* Local list
* Upvalue list


#### OpenDocument (.odt, .ods, .odp, .odb, .odg, .odf)

It is based on [XML] and [ZIP].


#### OpenRaster (.ora)

It is based on OpenDocument and [PNG].


#### PE file (.exe, .dll, .ocx, .scr, .cpl)

Leanify embedded resource.

Remove `Relocation Table` in executable file.

Remove undocumented `Rich Header`.

Overlap `PE Header` and `DOS Header`.


#### PNG image (.png, .apng)

Remove all ancillary chunks except for:

* `tRNS`: transparent information
* `fdAT`, `fcTL`, `acTL`: These chunks are used by `APNG`
* `npTc`: Android 9Patch images (*.9.png)

Optimize with `ZopfliPNG`.


#### RDB archive (.rdb)

It is an archive format found in Tencent QQ.

Leanify all files inside.


#### Flash file (.swf)

Leanify embedded images.

Recompress it with `LZMA`.

Remove Metadata Tag.


#### SVG image (.svg, .svgz)

It is based on [XML].

Remove metadata.

Shrink spaces in attributes.

Remove empty attributes.

Remove empty text element and container element.

#### tar archive (.tar)

Leanify all files inside.


#### XML document (.xml, .xsl, .xslt)

Remove all comments, unnecessary spaces, tabs, line breaks.


#### XPInstall (.xpi)

It is based on [ZIP].

Note that modifying files inside `xpi` will break digital signature.
To install it, you'll have to sign it again.


#### XPS document (.xps, .oxps)

It is based on [XML] and [ZIP].


#### ZIP archive (.zip)

Leanify all files inside and recompress deflate stream using [Zopfli](https://github.com/google/zopfli).

Use `STORE` method if `DEFLATE` makes file larger.

Remove extra field in `Local file header`.

Remove `Data descriptor structure`, write those information to `Local file header`.

Remove extra field and file comment in `Central directory file header`.

Remove comment in `End of central directory record`.



## Downloads

[Stable Releases](https://github.com/JayXon/Leanify/releases/)

[Windows Nightly Build](https://ci.appveyor.com/project/JayXon/leanify)



## Usage

```
Usage: leanify [options] paths
  -i, --iteration <iteration>   More iterations produce better result, but
                                  use more time, default is 15.
  -d, --max_depth <max depth>   Maximum recursive depth, unlimited by default.
                                  Set to 1 will disable recursive minifying.
  -f, --fastmode                Fast mode, no recompression.
  -q, --quiet                   No output to stdout.
  -v, --verbose                 Verbose output.
  --keep-exif                   Do not remove Exif.
```


## Compiling

#### Windows

* Visual Studio 2015+

  Use Leanify.vcxproj

* gcc 5+

  `build_gcc.bat` or `mingw32-make`


#### Linux, Mac

  gcc 5+ or clang 3.6+
```
make
```



[APK]: #apk-file-apk
[PNG]: #png-image-png-apng
[tar]: #tar-archive-tar
[XML]: #xml-document-xml-xsl-xslt
[ZIP]: #zip-archive-zip


CHANGELOG.md
# 0.4.4 (2021-03-10)

+ Added concurrent mode to utilize power of all CPUs

# 0.4.3 (2015-11-25)

+ Added support for Data URI #14.

+ Added support for DWF format.

+ Updated zopfli to 1.0.1.

+ Updated LZMA to 15.12.

+ Updated mozjpeg to 3.1.

+ Updated TinyXML2 to 3.0.0.

+ Improved PNG optimization for some cases.

+ Improved ZIP optimization for some cases.

+ Improved SVG minifying.

+ Added option `--keep-exif`.

+ Bug fixes and optimizations.


# 0.4.2 (2014-11-07)

+ Added support for PE.

+ Added support for tar.

+ Added support for fb2.

+ Added support for removing SVG metadata.

+ New -d option to specify maximum recursive depth.

+ Updated TinyXML2 to 2.2.0

+ Enable SSE2 and Link Time Optimization.

+ Win: Added 64-bit build.

+ Win: More meaningful error message.

+ Linux, Mac: Fixed printing UTF16 string.

+ Other optimizations and bug fixes.


# 0.4.1 (2014-08-10)

+ Add support for JPEG.

+ Add support for SWF.

+ Add support for XML with UTF8-BOM.

+ Improved command line arguments handling.

+ Fix gzip checksum.


# 0.4.0 (2014-08-04)

+ First public release.


main.cpp
#include "main.h"

#include <climits>
#include <cstring>
#include <iomanip>
#include <iostream>

#ifndef _WIN32
#include <ftw.h>
#endif

#include <taskflow/taskflow.hpp>

#include "fileio.h"
#include "formats/jpeg.h"
#include "formats/png.h"
#include "formats/zip.h"
#include "leanify.h"
#include "version.h"


using std::cerr;
using std::cout;
using std::endl;
using std::string;

template <typename T>
std::string ToString(const T a_value, const int n = 2) {
  std::ostringstream out;
  out.precision(n);
  out << std::fixed << a_value;
  return out.str();
}


const std::string BuildSize(size_t size) {
  if (size < 1024)
    return std::to_string(size) + " B";
  else if (size < 1024 * 1024)
    return ToString(size / 1024.0) + " KB";
  else
    return ToString(size / 1024.0 / 1024.0) + " MB";
}


#ifdef _WIN32
int ProcessFile(const std::wstring& file_path) {
  char mbs[MAX_PATH] = { 0 };
  WideCharToMultiByte(CP_ACP, 0, file_path.c_str(), -1, mbs, sizeof(mbs) - 1, nullptr, nullptr);
  string filename(mbs);
#else
int ProcessFile(const std::string& file_path) {
  const std::string& filename = file_path;
#endif  // _WIN32
  

  if (!parallel_processing)
    cout << "Processing: " << filename << endl;

  File input_file(file_path.c_str());

  if (input_file.IsOK()) {
    size_t original_size = input_file.GetSize();

    size_t new_size = LeanifyFile(input_file.GetFilePointer(), original_size, 0, filename);

    std::string log;
    if (parallel_processing)
      log = "Processed: " + filename + "\n";

    log += 
        BuildSize(original_size) + 
        " -> " + 
        BuildSize(new_size) +
        "\tLeanified: " + 
        BuildSize(original_size - new_size) + 
        " ("  + 
        ToString(100 - 100.0 * new_size / original_size) + 
        "%)";

    cout << log << endl;

    input_file.UnMapFile(new_size);
  }

  return 0;
}



void PauseIfNotTerminal() {
// pause if Leanify is not started in terminal
// so that user can see the output instead of just a flash of a black box
#ifdef _WIN32
  if (is_pause)
    system("pause");
#endif  // _WIN32
}


void PrintInfo() {
  cerr << "Leanify\t" << VERSION_STR << endl << endl;
  cerr << "Usage: leanify [options] paths\n"
          "  -i, --iteration <iteration>   More iterations may produce better result, but\n"
          "                                  use more time, default is 15.\n"
          "  -d, --max_depth <max depth>   Maximum recursive depth, unlimited by default.\n"
          "                                  Set to 1 will disable recursive minifying.\n"
          "  -f, --fastmode                Fast mode, no recompression.\n"
          "  -q, --quiet                   No output to stdout.\n"
          "  -v, --verbose                 Verbose output.\n"
          "  -p, --parallel                Distribute all tasks to all CPUs.\n"
          "  --keep-exif                   Do not remove Exif.\n"
          "  --keep-icc                    Do not remove ICC profile.\n"
          "\n"
          "JPEG options:\n"
          "  --jpeg-keep-all               Do not remove any metadata or comments in JPEG.\n"
          "  --jpeg-arithmetic             Use arithmetic coding for JPEG.\n"
          "\n"
          "ZIP options:\n"
          "  --zip-deflate                 Try deflate even if not compressed originally.\n";

  PauseIfNotTerminal();
}

tf::Taskflow taskflow;

#ifdef _WIN32
int EnqueueProcessFileTask(const wchar_t* file_path) {
  std::wstring* filePath = new std::wstring(file_path);
#else
// written like this in order to be callback function of ftw()
int EnqueueProcessFileTask(const char* file_path, const struct stat* sb = nullptr, int typeflag = FTW_F) {
  if (typeflag != FTW_F)
    return 0;
  std::string* filePath = new std::string(file_path);
#endif  // _WIN32

  auto task = [filePath]() { 
      ProcessFile(*filePath); 
      delete filePath;
  };
  taskflow.emplace(task);
  return 0;
}


#ifdef _WIN32
int main() {
  int argc;
  wchar_t** argv = CommandLineToArgvW(GetCommandLineW(), &argc);
#else
int main(int argc, char** argv) {
#endif  // _WIN32

  is_fast = false;
  is_verbose = false;
  iterations = 15;
  depth = 1;
  max_depth = INT_MAX;

#ifdef _WIN32
  is_pause = !getenv("PROMPT");
#endif  // _WIN32

  int i;
  for (i = 1; i < argc && argv[i][0] == L'-'; i++) {
#ifdef _WIN32
    // do not pause if any options are given
    is_pause = false;
#endif  // _WIN32
    int num_optargs = 0;
    for (int j = 1; argv[i][j]; j++) {
      switch (argv[i][j]) {
        case 'f':
          is_fast = true;
          break;
        case 'i':
          if (i < argc - 1) {
            iterations = STRTOL(argv[i + ++num_optargs], nullptr, 10);
            // strtol will return 0 on fail
            if (iterations == 0) {
              cerr << "There should be a positive number after -i option." << endl;
              PrintInfo();
              return 1;
            }
          }
          break;
        case 'd':
          if (i < argc - 1) {
            max_depth = STRTOL(argv[i + ++num_optargs], nullptr, 10);
            // strtol will return 0 on fail
            if (max_depth == 0) {
              cerr << "There should be a positive number after -d option." << endl;
              PrintInfo();
              return 1;
            }
          }
          break;
        case 'q':
          cout.setstate(std::ios::failbit);
          is_verbose = false;
          break;
        case 'v':
          cout.clear();
          is_verbose = true;
          break;
        case 'p':
          parallel_processing = true;
          break;
        case '-':
          if (STRCMP(argv[i] + j + 1, "fastmode") == 0) {
            j += 7;
            argv[i][j + 1] = 'f';
          } else if (STRCMP(argv[i] + j + 1, "iteration") == 0) {
            j += 8;
            argv[i][j + 1] = 'i';
          } else if (STRCMP(argv[i] + j + 1, "max_depth") == 0) {
            j += 8;
            argv[i][j + 1] = 'd';
          } else if (STRCMP(argv[i] + j + 1, "quiet") == 0) {
            j += 4;
            argv[i][j + 1] = 'q';
          } else if (STRCMP(argv[i] + j + 1, "verbose") == 0) {
            j += 6;
            argv[i][j + 1] = 'v';
          } else if (STRCMP(argv[i] + j + 1, "parallel") == 0) {
            j += 7;
            argv[i][j + 1] = 'p';
          } else if (STRCMP(argv[i] + j + 1, "keep-exif") == 0) {
            j += 9;
            Jpeg::keep_exif_ = true;
          } else if (STRCMP(argv[i] + j + 1, "keep-icc") == 0) {
            j += 8;
            Jpeg::keep_icc_profile_ = true;
            Png::keep_icc_profile_ = true;
          } else if (STRCMP(argv[i] + j + 1, "jpeg-keep-all") == 0) {
            j += 13;
            Jpeg::keep_all_metadata_ = true;
          } else if (STRCMP(argv[i] + j + 1, "jpeg-arithmetic") == 0) {
            j += 15;
            Jpeg::force_arithmetic_coding_ = true;
          } else if (STRCMP(argv[i] + j + 1, "zip-deflate") == 0) {
            j += 11;
            Zip::force_deflate_ = true;
          } else {
#ifdef _WIN32
            char mbs[64] = { 0 };
            WideCharToMultiByte(CP_ACP, 0, argv[i] + j + 1, -1, mbs, sizeof(mbs) - 1, nullptr, nullptr);
            cerr << "Unknown option: " << mbs << endl;
#else
            cerr << "Unknown option: " << argv[i] + j + 1 << endl;
#endif  // _WIN32
            PrintInfo();
            return 1;
          }
          break;
        default:
          cerr << "Unknown option: " << (char)argv[i][j] << endl;
          PrintInfo();
          return 1;
      }
    }
    i += num_optargs;
  }

  if (parallel_processing && is_verbose) {
    cerr << "Verbose logs not supported in parallel mode." << endl;
    PrintInfo();
    return 1;
  }

  if (i == argc) {
    cerr << "No file path provided." << endl;
    PrintInfo();
    return 1;
  }

  cout << std::fixed;
  cout.precision(2);

  // support multiple input file
  do {
    TraversePath(argv[i], EnqueueProcessFileTask);
  } while (++i < argc);

  size_t parallel_tasks = 1;
  if (parallel_processing)
    parallel_tasks = std::thread::hardware_concurrency();
  
  tf::Executor executor(parallel_tasks);
  executor.run(taskflow).wait(); 
  
  PauseIfNotTerminal();

  return 0;
}
